pyspectral/rayleigh.py

```diff
@@ -344,46 +304,66 @@ def reduce_rayleigh_highzenith(zenith, rayref, thresh_zen, maxzen, strength):
         be linearly scaled, from one at `thresh_zen` to zero at `maxzen`.
         """
         LOG.info("Reducing Rayleigh effect at high zenith angles.")
-        factor = 1. - strength * where(zenith < thresh_zen, 0, (zenith - thresh_zen) / (maxzen - thresh_zen))
+        factor = 1. - strength * np.where(zenith < thresh_zen, 0, (zenith - thresh_zen) / (maxzen - thresh_zen))
         # For low zenith factor can be greater than one, so we need to clip it into a sensible range.
-        factor = clip(factor, 0, 1)
+        factor = np.clip(factor, 0, 1)
         return rayref * factor
 
 
-def get_reflectance_lut_from_file(filename):
+def _get_zeroed_result(shape, compute, chunks=None):
+    if HAVE_DASK:
+        res = da.zeros(shape, chunks=chunks)
+        return res.compute() if compute else res
+    return np.zeros(shape)
+
+
+def _get_wavelength_adjusted_lut_rayleigh_reflectance(lut_filename, wvl):
+    rayleigh_refl, wvl_coord = _get_rayleigh_wavelength_lut_from_file(lut_filename)
+    wvl_coord = wvl_coord[:]
+    if not wvl_coord.min() < wvl < wvl_coord.max():
+        return None
+    wavelength_index, wavelength_factor = _get_wavelength_index_and_factor(wvl_coord, wvl)
+    raylwvl = (wavelength_factor * rayleigh_refl[wavelength_index - 1, :, :, :] +
+               (1 - wavelength_factor) * rayleigh_refl[wavelength_index, :, :, :])
+    return raylwvl
+
+
+def _get_wavelength_index_and_factor(wvl_coord, wvl):
+    wavelength_index = np.searchsorted(wvl_coord, wvl)
+    wvl1 = wvl_coord[wavelength_index - 1]
+    wvl2 = wvl_coord[wavelength_index]
+    wavelength_factor = (wvl2 - wvl) / (wvl2 - wvl1)
+    return wavelength_index, wavelength_factor
+
+
+def _get_rayleigh_wavelength_lut_from_file(lut_filename):
+    h5f = h5py.File(lut_filename, 'r')
+    rayleigh_refl = h5f["reflectance"]
+    wvl_coord = h5f["wavelengths"]
+    return rayleigh_refl, wvl_coord
+
+
+def get_reflectance_lut_from_file(lut_filename):
     """Get reflectance LUT.
 
     Read the Look-Up Tables from file with reflectances as a function of
     wavelength, satellite zenith secant, azimuth difference angle, and sun
-    zenith secant
+    zenith secant.
 
     """
-    h5f = h5py.File(filename, 'r')
+    h5f = h5py.File(lut_filename, 'r')
```

Comment:
I think it would be cleaner and more exception-proof if we used a context manager here:
```python
with h5py.File(lut_filename, 'r') as h5f:
   ...

```

Reply:
So it could have been a fluke, but when I did this on the `_get_rayleigh_wavelength_lut_from_file` function, it crashed with some weird error. Probably because I'm returning the HDF5 objects and by the time they are used the file is closed. I'll double check though.

Reply:
This has been done now.